lib/launcher/main.c

```diff
@@ -0,0 +1,353 @@
+#include "cuda.h"
+#include <Python.h>
+
+union ArgUnion {
+  PyObject *O;
+  float f;
+  double d;
+  long l;
+  uint32_t I;
+  int32_t i;
+  uint64_t K;
+  int64_t L;
+  CUdeviceptr CUdptr;
+};
+
+static inline void gpuAssert(CUresult code, const char *file, int line) {
+  if (code != CUDA_SUCCESS) {
+    const char *prefix = "Triton Error [CUDA]: ";
+    const char *str;
+    cuGetErrorString(code, &str);
+    char err[1024] = {0};
+    strcat(err, prefix);
+    strcat(err, str);
+    PyErr_SetString(PyExc_RuntimeError, err);
+  }
+}
+#define CUDA_CHECK(ans)                                                        \
+  { gpuAssert((ans), __FILE__, __LINE__); }
+
+void my_launch(int gridX, int gridY, int gridZ, int num_warps,
+               int shared_memory, CUstream stream, CUfunction function,
+               void **params) {
+  if (gridX * gridY * gridZ > 0) {
+    CUDA_CHECK(cuLaunchKernel(function, gridX, gridY, gridZ, 32 * num_warps, 1,
+                              1, shared_memory, stream, params, 0));
+  }
+}
+
+static inline CUdeviceptr getPointer(PyObject *obj, int idx) {
+  if (PyLong_Check(obj)) {
+    return (CUdeviceptr)PyLong_AsUnsignedLongLong(obj);
+  }
+  if (obj == Py_None) {
+    return (CUdeviceptr)0;
+  }
+  PyObject *ptr = PyObject_GetAttrString(obj, "data_ptr");
+  if (ptr) {
+    PyObject *empty_tuple = PyTuple_New(0);
+    PyObject *ret = PyObject_Call(ptr, empty_tuple, NULL);
+    Py_DECREF(empty_tuple);
+    Py_DECREF(ptr);
+    if (!PyLong_Check(ret)) {
+      PyErr_SetString(
+          PyExc_TypeError,
+          "data_ptr method of Pointer object must return 64-bit int");
+    }
+    return (CUdeviceptr)PyLong_AsUnsignedLongLong(ret);
+  }
+  PyErr_SetString(
+      PyExc_TypeError,
+      "Pointer argument must be either uint64 or have data_ptr method");
+  return (CUdeviceptr)0;
+}
+
+static PyObject *launch(PyObject *self, PyObject *args) {
+  char *format = (char *)PyUnicode_AsUTF8(PyTuple_GetItem(args, 0));
+  char *is_const = (char *)PyUnicode_AsUTF8(PyTuple_GetItem(args, 1));
+  int gridX, gridY, gridZ;
+  uint64_t _stream;
+  uint64_t _function;
+  int num_warps;
+  int shared_memory;
+  PyObject *launch_enter_hook = NULL;
+  PyObject *launch_exit_hook = NULL;
+  PyObject *compiled_kernel = NULL;
+  PyObject *hook_ret = NULL;
+
+  int ArgC = strlen(format) - strlen("ssiiiiiKKOOO");
+  union ArgUnion MyArgs[100];
+
+  if (0)
```

Comment:
What's the purpose of this `if`?

Reply:
(not ready for review, some traces of debugging are left / i didn't try to clean up this code cause was unsure if it's of any interest) (but huge thanks for the quick reply - I'll update my PR soon)

Reply:
Got you. Sorry I thought it's ready since you've changed it from draft to ready for review.